src/algebra/category/FinVect.lean

```diff
@@ -0,0 +1,167 @@
+
+protected lemma coe_comp {U V W : FinVect K} (f : U ⟶ V) (g : V ⟶ W) :
+  ((f ≫ g) : U → W) = (g : V → W) ∘ (f : U → V) := rfl
+
+instance monoidal_category : monoidal_category (FinVect K) :=
+monoidal_category.full_monoidal_subcategory
+  (λ V, finite_dimensional K V)
+  (finite_dimensional.finite_dimensional_self K)
+  (λ X Y hX hY, by exactI finite_dimensional_tensor_product X Y)
+
+variables (V : FinVect K)
+
+/-- The dual module is the dual in the rigid monoidal category `FinVect K`. -/
+def FinVect_dual : FinVect K :=
+⟨Module.of K (module.dual K V), subspace.module.dual.finite_dimensional⟩
+
+instance : has_coe_to_fun (FinVect_dual K V) :=
+{ F := λ v, V → K,
+  coe := λ v, by { change V →ₗ[K] K at v, exact v, }, }
+
+open category_theory.monoidal_category
+
+/-- The coevaluation map is defined in `linear_algebra.coevaluation`. -/
+def FinVect_coevaluation : 𝟙_ (FinVect K) ⟶ V ⊗ (FinVect_dual K V) :=
+by apply coevaluation K V
+
+lemma FinVect_coevaluation_apply_one : FinVect_coevaluation K V (1 : K) =
+   ∑ (i : basis.of_vector_space_index K V),
+    (basis.of_vector_space K V) i ⊗ₜ[K] (basis.of_vector_space K V).coord i :=
+by apply coevaluation_apply_one K V
+
+/-- The evaluation morphism is given by the contraction map. -/
+def FinVect_evaluation : (FinVect_dual K V) ⊗ V ⟶ 𝟙_ (FinVect K) :=
+by apply contract_left K V
+
+lemma FinVect_evaluation_apply (f : (FinVect_dual K V)) (x : V) :
+  (FinVect_evaluation K V) (f ⊗ₜ x) = f x :=
+by apply contract_left_apply f x
+
+@[simp]
+lemma right_unitor_hom_apply_tensor_one (x : V) :
+  ((ρ_ V).hom : _ →ₗ[K] _) (x ⊗ₜ[K] (1 : K)) = x :=
+(right_unitor_hom_apply x 1).trans (one_smul _ _)
+
+@[simp]
+lemma left_unitor_hom_apply_one_tensor (x : V) :
+  ((λ_ V).hom : _ →ₗ[K] _) ((1 : K) ⊗ₜ[K] x) = x :=
+(left_unitor_hom_apply 1 x).trans (one_smul _ _)
+
+@[simp]
+lemma left_unitor_inv_apply (x : V) : ((λ_ V).inv : _ →ₗ[K] _) x = (1 : K) ⊗ₜ[K] x :=
+left_unitor_inv_apply _
+
+@[simp]
+lemma right_unitor_inv_apply (x : V) : ((ρ_ V).inv : _ →ₗ[K] _) x = x ⊗ₜ[K] (1 : K) :=
+right_unitor_inv_apply _
+
+@[simp]
+lemma tensor_hom_apply {U V W X : FinVect K} (f : U ⟶ V) (g : W ⟶ X) (k : U) (m : W) :
+  ((f ⊗ g) : U ⊗ W ⟶ V ⊗ X) (k ⊗ₜ m) = f k ⊗ₜ g m :=
+hom_apply f g k m
+
+@[simp]
+lemma id_apply {V : FinVect K} (x : V) : (𝟙 V : _ →ₗ[K] _) x = x := rfl
+
+@[simp]
+lemma associator_inv_apply {U V W : FinVect K} (u : U) (v : V) (w : W) :
+  ((α_ U V W).inv : U ⊗ (V ⊗ W) ⟶ (U ⊗ V) ⊗ W) (u ⊗ₜ (v ⊗ₜ w)) = ((u ⊗ₜ v) ⊗ₜ w) :=
+associator_inv_apply u v w
+
+@[simp]
+lemma associator_hom_apply {U V W : FinVect K} (u : U) (v : V) (w : W) :
+  ((α_ U V W).hom : (U ⊗ V) ⊗ W ⟶ U ⊗ (V ⊗ W)) ((u ⊗ₜ v) ⊗ₜ w) = (u ⊗ₜ (v ⊗ₜ w)) :=
+associator_hom_apply u v w
+
+private theorem coevaluation_evaluation :
+  let V' : FinVect K := FinVect_dual K V in
+  (𝟙 V' ⊗ (FinVect_coevaluation K V)) ≫ (α_ V' V V').inv ≫ (FinVect_evaluation K V ⊗ 𝟙 V')
+  = (ρ_ V').hom ≫ (λ_ V').inv :=
```

Comment:
How hard do you think it would be to prove the plain-old-linear-algebra statement of this fact first, over in `linear_algebra.coevaluation`, and refactor this proof in terms of it? (Ditto for the next statement.)

I don't think it's essential, but it might be nicer that way.

Reply:
I'll give it a try!

I can't really judge, is the fact actually interesting outside of establishing the dual structure? Why put it in `linear_algebra.coevaluation` as opposed to `linear_algebra.contraction`?

Reply:
This is the statement in `linear_algebra` terms (is this or the `ext`ed version preferred?):
```lean
lemma coevaluation_evaluation (x : K) (v : module.dual K V) :
  (((contract_left K V).rtensor _) $
  (tensor_product.assoc K _ _ _).symm $
  ((coevaluation K V).ltensor _) (v ⊗ₜ[K] x))
  = (tensor_product.lid K _).symm (tensor_product.rid K _ (v ⊗ₜ[K] x)) :=
sorry
```

Reply:
I transferred the lemmas over, that makes `FinVect` neat and short!